FPGA: Add sample for `max_reinvocation_delay` attribute (#5)
williamwang256/oneAPI-samples: DirectProgramming/C++SYCL_FPGA/Tutorials/Features/max_reinvocation_delay/README.md

## Prerequisites

| Optimized for        | Description
|:---                  |:---
| OS                   | Ubuntu* 18.04/20.04 <br> RHEL*/CentOS* 8 <br> SUSE* 15 <br> Windows* 10
| Hardware             | Intel® Agilex® 7, Arria® 10, and Stratix® 10 FPGAs
| Software             | Intel® oneAPI DPC++/C++ Compiler

> **Note**: Even though the Intel® oneAPI DPC++/C++ Compiler is enough to compile for emulation, generating reports and generating RTL, there are extra software requirements for the simulation flow and FPGA compiles.
>
> For using the simulator flow, Intel® Quartus® Prime Pro Edition and one of the following simulators must be installed and accessible through your PATH:
> - Questa*-Intel® FPGA Edition
> - Questa*-Intel® FPGA Starter Edition
> - ModelSim® SE
>
> When using the hardware compile flow, Intel® Quartus® Prime Pro Edition must be installed and accessible through your PATH.

> **Warning**: Make sure you add the device files associated with the FPGA that you are targeting to your Intel® Quartus® Prime installation.

> This sample is part of the FPGA code samples.

Reviewer: Make sure you register your sample in the top level FPGA readme

   style tier2 fill:#f96,stroke:#333,stroke-width:1px,color:#fff
   style tier3 fill:#0071c1,stroke:#0071c1,stroke-width:1px,color:#fff
   style tier4 fill:#0071c1,stroke:#0071c1,stroke-width:1px,color:#fff
```

Find more information about how to navigate this part of the code samples in the [FPGA top-level README.md](/DirectProgramming/C++SYCL_FPGA/README.md).
You can also find more information about [troubleshooting build errors](/DirectProgramming/C++SYCL_FPGA/README.md#troubleshooting), [running the sample on the Intel® DevCloud](/DirectProgramming/C++SYCL_FPGA/README.md#build-and-run-the-samples-on-intel-devcloud-optional), [using Visual Studio Code with the code samples](/DirectProgramming/C++SYCL_FPGA/README.md#use-visual-studio-code-vs-code-optional), [links to selected documentation](/DirectProgramming/C++SYCL_FPGA/README.md#documentation), and more.

## Key Implementation Details

The sample illustrates the following important concepts.

- How to use the `max_reinvocation_delay` attribute to restrict the maximum delay between loop invocations.

### Description of the `max_reinvocation_delay` Attribute

> Apply the `[[intel::max_reinvocation_delay(N)]]` attribute to a loop to restrict the delay between invocations of that loop. The attribute parameter `N` is required and must be a positive constant expression of integer type (currently, only `N=1` is supported, i.e., that there shall be no delay between invocations). This parameter specifies the maximum number of clock cycles allowed between the last iteration of a loop invocation and the first iteration of the next invocation of the loop. The higher the maximum reinvocation delay allowed, the longer the wait before the next loop invocation can start executing.

Reviewer: Why is the max reinvocation delay not set to the minimum possible value by the compiler to begin with?
As an external user I would think of this max_reinvocation_delay feature to be a "quick fix" for a compiler inefficiency

So I'm guessing there is a reason why it's not the case by default, and I think we should make it clear here.

Author: updated! according to the documentation:

"the compiler can optimize the loop better if a high reinvocation delay is allowed in one of the following ways: Enable loop speculation; Pipeline the loop orchestration hardware that can increase fMAX".

# `max_reinvocation_delay` Sample

This sample is an FPGA tutorial that explains how to use the `max_reinvocation_delay` attribute for loops.

| Area                 | Description
|:--                   |:--

> | What you will learn  | How and when to apply the `max_reinvocation_delay` attribute to restrict the maximum delay between loop invocations.

Reviewer 2: ```suggestion
| What you will learn  | How and when to apply the `max_reinvocation_delay` attribute when optimizing loop throughput
```

## Prerequisites

| Optimized for        | Description
|:---                  |:---
| OS                   | Ubuntu* 18.04/20.04 <br> RHEL*/CentOS* 8 <br> SUSE* 15 <br> Windows* 10
| Hardware             | Intel® Agilex® 7, Arria® 10, and Stratix® 10 FPGAs
| Software             | Intel® oneAPI DPC++/C++ Compiler

> **Note**: Even though the Intel® oneAPI DPC++/C++ Compiler is enough to compile for emulation, generating reports and generating RTL, there are extra software requirements for the simulation flow and FPGA compiles.
>
> For using the simulator flow, Intel® Quartus® Prime Pro Edition and one of the following simulators must be installed and accessible through your PATH:
> - Questa*-Intel® FPGA Edition
> - Questa*-Intel® FPGA Starter Edition
> - ModelSim® SE
>
> When using the hardware compile flow, Intel® Quartus® Prime Pro Edition must be installed and accessible through your PATH.

> **Warning**: Make sure you add the device files associated with the FPGA that you are targeting to your Intel® Quartus® Prime installation.

This sample is part of the FPGA code samples.

> It is categorized as a Tier 2 sample that demonstrates a compiler feature.

Reviewer 2: ```suggestion
It is categorized as a Tier 3 sample that [whatever should go here].
```

   tier1("Tier 1: Get Started")
   tier2("Tier 2: Explore the Fundamentals")
   tier3("Tier 3: Explore the Advanced Techniques")
   tier4("Tier 4: Explore the Reference Designs")

   tier1 --> tier2 --> tier3 --> tier4

   style tier1 fill:#0071c1,stroke:#0071c1,stroke-width:1px,color:#fff
   style tier2 fill:#f96,stroke:#333,stroke-width:1px,color:#fff
   style tier3 fill:#0071c1,stroke:#0071c1,stroke-width:1px,color:#fff
   style tier4 fill:#0071c1,stroke:#0071c1,stroke-width:1px,color:#fff
```

Find more information about how to navigate this part of the code samples in the [FPGA top-level README.md](/DirectProgramming/C++SYCL_FPGA/README.md).
You can also find more information about [troubleshooting build errors](/DirectProgramming/C++SYCL_FPGA/README.md#troubleshooting), [running the sample on the Intel® DevCloud](/DirectProgramming/C++SYCL_FPGA/README.md#build-and-run-the-samples-on-intel-devcloud-optional), [using Visual Studio Code with the code samples](/DirectProgramming/C++SYCL_FPGA/README.md#use-visual-studio-code-vs-code-optional), [links to selected documentation](/DirectProgramming/C++SYCL_FPGA/README.md#documentation), and more.

## Key Implementation Details

> The sample illustrates the following important concepts.
> 
> - How and when to apply the `max_reinvocation_delay` attribute to restrict the maximum delay between loop invocations.
> 
> ### Description of the `max_reinvocation_delay` Attribute
> 

Reviewer 2: ```suggestion
```
don't need this additional level of heirarchy.

Apply the `[[intel::max_reinvocation_delay(N)]]` attribute to loops in your program on which you want to specify a maximum loop reinvocation delay. The *loop reinvocation delay* is defined as the latency between the last iteration of a loop invocation and the first iteration of the next invocation of that loop. 

> > **Note:** For interleaved loops, the loop reinvocation delay is defined as the latency between the last iteration of a loop invocation and the first iteration of the next loop invocation immediately following it, which may not be the next loop invocation in the program order.

Reviewer 2: this feels like a repetition of the immediately previous line. one or the other is enough :)

> 

Reviewer 2: ```suggestion
> **Note**:  A loop **invocation** begins when the program flow enters a loop for the first time, while an **iteration** begins each time the program flow enters the loop body. In the example below, the `i` loop will have `sequence_length` iterations for each invocation. It will be invoked once for each **iteration** of the outer `factor` loop.
```
might be worth it to clear up the difference between iterations and invocations.

> The attribute parameter `N` is required and must be a positive constant expression of integer type. This parameter controls the maximum loop reinvocation delay allowed, measured in clock cycles. The higher the maximum reinvocation delay allowed, the longer the wait before the next loop invocation can start executing. Currently, only `N=1` is supported, i.e., that there should be no delay between invocations.

Reviewer 2: ```suggestion
The attribute parameter `N` is required and must be a positive constant expression of integer type. This parameter controls the maximum loop reinvocation delay allowed, measured in clock cycles. Currently, only `N=1` is supported, i.e., that there should be no delay between invocations. If you do not specify this attribute, the compiler may insert some delay to improve fMAX as shown in the following example.
```

no need to mention what happens if you give it a bigger number as that isn't supported.

> #### Example

Reviewer 2: ```suggestion
### Example
```
up a level of heirarchy

> Consider the following nested loop:

Reviewer 2: ```suggestion
Consider the following loop nest:
```

```c++
for (int factor = 0; factor < FACTORS; factor++) {
  for (int i = 0; i < sequence_length; i++) {
    PipeOut::write(first_term + i * factor);
  }
}
```

> By default, the compiler inserts a loop reinvocation delay of a few cycles on the inner loop. In general, the compiler may insert a higher loop reinvocation delay to better optimize a loop by enabling loop speculation or by pipelining the loop orchestration hardware, which can increase fMAX.

Reviewer 2: ```suggestion
By default, the compiler inserts a loop reinvocation delay of a few cycles on the inner loop. In general, the compiler may insert a higher loop reinvocation delay to better optimize a loop by enabling loop speculation or by pipelining the loop orchestration hardware, which can increase fMAX. This delay can be seen by observing the transactions coming out of the streaming interface associated with `PipeOut`.
```

![default behavior with delay between invocations](assets/default.png)

However, the extra latency between invocations of a loop can have a significant impact in performance if the loop has a very small trip count. Applying `[[intel::max_reinvocation_delay(1)]]` to the inner loop allows us to remove this delay between invocations. 

```c++
for (int factor = 0; factor < FACTORS; factor++) {
  [[intel::max_reinvocation_delay(1)]]
  for (int i = 0; i < sequence_length; i++) {
    PipeOut::write(first_term + i * factor);
  }
}
```

> Now, the first iteration of the `i + 1` st invocation of the inner loop will launch immediately following the last iteration of the `i` th invocation of the inner loop. 

Reviewer 2: ```suggestion
Now, the first iteration of the `i+1`th  invocation of the inner loop will launch immediately following the final iteration of the `i`th invocation of the inner loop. 
```

> - For non-POSIX shells, like csh, use the following command: `bash -c 'source <install-dir>/setvars.sh ; exec csh'`
>
> Windows*:
> - `C:\Program Files(x86)\Intel\oneAPI\setvars.bat`
> - Windows PowerShell*, use the following command: `cmd.exe "/K" '"C:\Program Files (x86)\Intel\oneAPI\setvars.bat" && powershell'`
>
> For more information on configuring environment variables, see [Use the setvars Script with Linux* or macOS*](https://www.intel.com/content/www/us/en/develop/documentation/oneapi-programming-guide/top/oneapi-development-environment-setup/use-the-setvars-script-with-linux-or-macos.html) or [Use the setvars Script with Windows*](https://www.intel.com/content/www/us/en/develop/documentation/oneapi-programming-guide/top/oneapi-development-environment-setup/use-the-setvars-script-with-windows.html).

### On Linux*

1. Change to the sample directory.
2. Build the program for Intel® Agilex® 7 device family, which is the default.
   ```
   mkdir build
   cd build
   cmake ..
   ```
   > **Note**: You can change the default target by using the command:
   >  ```
   >  cmake .. -DFPGA_DEVICE=<FPGA device family or FPGA part number>
   >  ```

>    > This tutorial only uses the IP Authoring flow and does not support targeting an explicit FPGA board variant and BSP.

Reviewer 2: ```suggestion
   > For simplicity, this tutorial only uses the IP Authoring flow and does not support targeting an explicit FPGA board variant and BSP.
```

3. Compile the design. (The provided targets match the recommended development flow.)

   1. Compile and run for emulation (fast compile time, targets emulated FPGA device).
      ```
      make fpga_emu
      ```
   2. Generate the optimization reports.
      ```
      make report
      ```
   3. Compile and run for simulation (fast compile time, targets simulated FPGA device).
      ```
      make fpga_sim
      ```
   4. Compile for FPGA hardware (longer compile time, targets an FPGA device).
      ```
      make fpga
      ```

### On Windows*

1. Change to the sample directory.
2. Build the program for the Intel® Agilex® 7 device family, which is the default.
   ```
   mkdir build
   cd build
   cmake -G "NMake Makefiles" ..
   ```
   > **Note**: You can change the default target by using the command:
   >  ```
   >  cmake -G "NMake Makefiles" .. -DFPGA_DEVICE=<FPGA device family or FPGA part number>
   >  ```

>    > This tutorial only uses the IP Authoring flow and does not support targeting an explicit FPGA board variant and BSP.

Reviewer 2: ```suggestion
   > For simplicity, this tutorial only uses the IP Authoring flow and does not support targeting an explicit FPGA board variant and BSP.
```

   style tier2 fill:#f96,stroke:#333,stroke-width:1px,color:#fff
   style tier3 fill:#0071c1,stroke:#0071c1,stroke-width:1px,color:#fff
   style tier4 fill:#0071c1,stroke:#0071c1,stroke-width:1px,color:#fff
```

Find more information about how to navigate this part of the code samples in the [FPGA top-level README.md](/DirectProgramming/C++SYCL_FPGA/README.md).
You can also find more information about [troubleshooting build errors](/DirectProgramming/C++SYCL_FPGA/README.md#troubleshooting), [running the sample on the Intel® DevCloud](/DirectProgramming/C++SYCL_FPGA/README.md#build-and-run-the-samples-on-intel-devcloud-optional), [using Visual Studio Code with the code samples](/DirectProgramming/C++SYCL_FPGA/README.md#use-visual-studio-code-vs-code-optional), [links to selected documentation](/DirectProgramming/C++SYCL_FPGA/README.md#documentation), and more.

## Key Implementation Details

The sample illustrates the following important concepts.

- How to use the `max_reinvocation_delay` attribute to restrict the maximum delay between loop invocations.

### Description of the `max_reinvocation_delay` Attribute

> Apply the `[[intel::max_reinvocation_delay(N)]]` attribute to a loop to restrict the delay between invocations of that loop. The attribute parameter `N` is required and must be a positive constant expression of integer type (currently, only `N=1` is supported). This parameter specifies the maximum number of clock cycles allowed between the last iteration of a loop invocation and the first iteration of the next invocation of the loop. The higher the maximum reinvocation delay allowed, the longer the wait before the next loop invocation can start executing.

Reviewer 3: maybe explain when N=1, it means no delay.

## Key Implementation Details

Apply the `[[intel::max_reinvocation_delay(N)]]` attribute to loops in your program on which you want to specify a maximum loop reinvocation delay. The *loop reinvocation delay* is defined as the latency between the last iteration of a loop invocation and the first iteration of the next invocation of that loop. 

> **Note:** For interleaved loops, the loop reinvocation delay is defined as the latency between the last iteration of a loop invocation and the first iteration of the next loop invocation immediately following it, which may not be the next loop invocation in the program order.
> **Note**:  A loop **invocation** begins when the program flow enters a loop for the first time, while an **iteration** begins each time the program flow enters the loop body. In the example below, the `i` loop will have `sequence_length` iterations for each invocation. It will be invoked once for each **iteration** of the outer `factor` loop.

> The attribute parameter `N` is required and must be a positive constant expression of integer type. This parameter controls the maximum loop reinvocation delay allowed, measured in clock cycles. Currently, only `N=1` is supported, i.e., that there should be no delay between invocations. If you do not specify this attribute, the compiler may insert some delay to improve fMAX as shown in the following example.

Reviewer 2: ```suggestion
Apply the `[[intel::max_reinvocation_delay(N)]]` attribute to loops in your program on which you want to specify a maximum loop reinvocation delay. The *loop reinvocation delay* is defined as the latency between the last iteration of a loop invocation and the first iteration of the next invocation of that loop. 

> **Note:** For *interleaved* loops, the loop reinvocation delay is defined as the latency between the last iteration of a loop invocation and the first iteration of the next loop invocation immediately following it, which may not be the next loop invocation in the program order.

> **Note**:  A loop **invocation** begins when the program flow enters a loop for the first time, while an **iteration** begins each time the program flow enters the loop body. In the example below, the `i` loop will have `sequence_length` iterations for each invocation. It will be invoked once for each **iteration** of the outer `factor` loop.

The attribute parameter `N` is required and must be a positive constant expression of integer type. This parameter controls the maximum loop reinvocation delay allowed, measured in clock cycles. Currently, only `N=1` is supported, that is, that there should be no delay between invocations. If you do not specify this attribute, the compiler may insert some delay to improve fMAX as shown in the following example.
```

need to leave empty lines to get markdown to make a line break

Author: done in separate commit
